server/src/main/java/org/apache/druid/messages/server/OutboxImpl.java

```diff
@@ -0,0 +1,226 @@
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Production implementation of {@link Outbox}. Each outbox is represented by an {@link OutboxQueue}.
+ */
+public class OutboxImpl<MessageType> implements Outbox<MessageType>
+{
+  private static final int MAX_BATCH_SIZE = 8;
+
+  // clientHost -> outgoing message queue
+  private final ConcurrentHashMap<String, OutboxQueue<MessageType>> queues;
+  private volatile boolean stopped;
+
+  public OutboxImpl()
+  {
+    this.queues = new ConcurrentHashMap<>();
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    stopped = true;
+
+    final Iterator<OutboxQueue<MessageType>> it = queues.values().iterator();
+    while (it.hasNext()) {
+      it.next().stop();
+      it.remove();
+    }
+  }
+
+  @Override
+  public ListenableFuture<?> sendMessage(String clientHost, MessageType message)
+  {
+    if (stopped) {
+      return Futures.immediateCancelledFuture();
+    }
+
+    return queues.computeIfAbsent(clientHost, id -> new OutboxQueue<>())
+                 .sendMessage(message);
+  }
+
+  @Override
+  public ListenableFuture<MessageBatch<MessageType>> getMessages(String clientHost, long epoch, long startWatermark)
+  {
+    if (stopped) {
+      return Futures.immediateCancelledFuture();
+    }
+
+    final OutboxQueue<MessageType> queue = queues.computeIfAbsent(clientHost, id -> new OutboxQueue<>());
+    if (epoch != queue.epoch && epoch != MessageRelay.INIT) {
+      return Futures.immediateFuture(new MessageBatch<>(Collections.emptyList(), queue.epoch, 0));
+    }
+
+    return queue.getMessages(startWatermark);
+  }
+
+  @Override
+  public void resetOutbox(final String clientHost)
+  {
+    final OutboxQueue<MessageType> queue = queues.remove(clientHost);
+    if (queue != null) {
+      queue.stop();
+    }
+  }
+
+  @VisibleForTesting
+  long getOutboxEpoch(final String clientHost)
+  {
+    final OutboxQueue<MessageType> queue = queues.get(clientHost);
+    return queue != null ? queue.epoch : MessageRelay.INIT;
+  }
+
+  /**
+   * Outgoing queue for a specific client.
+   */
+  public static class OutboxQueue<T>
+  {
+    /**
+     * Epoch, set when the outbox is created. Attached to returned batches through {@link MessageBatch#getEpoch()}.
+     */
+    private final long epoch;
+
+    /**
+     * Currently-outstanding futures. These are tracked so they can be canceled in {@link #stop()}.
+     */
+    private final Set<ListenableFuture<?>> pendingFutures = Sets.newConcurrentHashSet();
+
+    @GuardedBy("this")
+    private long startWatermark = 0;
+
+    @GuardedBy("this")
+    private final Deque<Pair<SettableFuture<?>, T>> queue = new ArrayDeque<>();
+
+    @GuardedBy("this")
+    private SettableFuture<?> messageAvailableFuture = SettableFuture.create();
+
+    private volatile boolean stopped;
+
+    public OutboxQueue()
+    {
+      this.epoch = ThreadLocalRandom.current().nextLong() & Long.MAX_VALUE;
```

Comment:
I wonder if its simpler to have an auto-incrementing counter here instead?

Reply:
This also needs to be different from a previous version from the same host after a JVM reboot, so it can't simply start at zero. I figured a random 63 bit int would be sufficient.